src/_image_wrapper.cpp

```diff
@@ -49,249 +50,177 @@ const char* image_resample__doc__ =
+
+    if (out_ndim == 3 && output_array.shape(2) != 4)
+        throw std::invalid_argument(
+            "3D output array must be RGBA with shape (M, N, 4), has trailing dimension of " +
+            std::to_string(out_ndim));
 
-    if (py_transform == NULL || py_transform == Py_None) {
+    if (!output_array.dtype().is(dtype))
+        throw std::invalid_argument("Input and output arrays have mismatched types");
+
+    if ((output_array.flags() & pybind11::array::c_style) == 0)
+        throw std::invalid_argument("Output array must be C-contiguous");
+
+    if (!output_array.writeable())
+        throw std::invalid_argument("Output array must be writeable");
+
+    resample_params_t params;
+    params.interpolation = interpolation;
+    params.transform_mesh = nullptr;
+    params.resample = resample_;
+    params.norm = norm;
+    params.radius = radius;
+    params.alpha = alpha;
+
+    // Only used if transform is not affine.
+    // Need to keep it in scope for the duration of this function.
+    pybind11::array_t<double> transform_mesh;
+
+    // Validate transform
+    if (transform.is_none()) {
         params.is_affine = true;
     } else {
-        PyObject *py_is_affine;
-        int py_is_affine2;
-        py_is_affine = PyObject_GetAttrString(py_transform, "is_affine");
-        if (!py_is_affine) {
-            goto error;
-        }
+        // Raises Python AttributeError if no such attribute or TypeError if cast fails
+        bool is_affine = pybind11::cast<bool>(transform.attr("is_affine"));
 
-        py_is_affine2 = PyObject_IsTrue(py_is_affine);
-        Py_DECREF(py_is_affine);
-
-        if (py_is_affine2 == -1) {
-            goto error;
-        } else if (py_is_affine2) {
-            if (!convert_trans_affine(py_transform, &params.affine)) {
-                goto error;
-            }
+        if (is_affine) {
+            convert_trans_affine(transform, params.affine);
             params.is_affine = true;
         } else {
-            transform_mesh = _get_transform_mesh(
-                py_transform, PyArray_DIMS(output));
-            if (!transform_mesh) {
-                goto error;
-            }
-            params.transform_mesh = (double *)PyArray_DATA(transform_mesh);
+            transform_mesh = _get_transform_mesh(transform, output_array.shape());
+            params.transform_mesh = transform_mesh.data();
             params.is_affine = false;
         }
     }
 
     if (auto resampler =
             (ndim == 2) ? (
-                (type == NPY_UINT8) ? resample<agg::gray8> :
-                (type == NPY_INT8) ? resample<agg::gray8> :
-                (type == NPY_UINT16) ? resample<agg::gray16> :
-                (type == NPY_INT16) ? resample<agg::gray16> :
-                (type == NPY_FLOAT32) ? resample<agg::gray32> :
-                (type == NPY_FLOAT64) ? resample<agg::gray64> :
+                (dtype.is(pybind11::dtype::of<std::uint8_t>())) ? resample<agg::gray8> :
```

Comment:
This is more verbose than it used to be, but is the recommended way of checking `dtype`s in `pybind11`.